x-pack/plugins/fleet/server/services/epm/package_service.ts

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { KibanaRequest, ElasticsearchClient, SavedObjectsClient, Logger } from 'kibana/server';
+import type {
+  EsAssetReference,
+  InstallablePackage,
+  RegistryPackage,
+  RegistrySearchResult,
+} from '../../types';
+import { ElasticsearchAssetType } from '../../types';
+import { checkSuperuser } from '../../routes/security';
+import { FleetUnauthorizedError } from '../../errors';
+import { getPathParts } from './archive';
+import { installTransform } from './elasticsearch/transform/install';
+import { fetchFindLatestPackage, getRegistryPackage } from './registry';
+
+export type InstalledAssetType = EsAssetReference;
+
+export interface PackageService {
+  asScoped(request: KibanaRequest): PackageClient;
+  asInternalUser: PackageClient;
+}
+
+export interface PackageClient {
+  fetchFindLatestPackage(packageName: string): Promise<RegistrySearchResult>;
+
+  getRegistryPackage(
+    packageName: string,
+    packageVersion: string
+  ): Promise<{ packageInfo: RegistryPackage; paths: string[] }>;
+
+  reinstallEsAssets(
+    assetType: ElasticsearchAssetType,
+    packageInfo: InstallablePackage,
+    paths: string[]
+  ): Promise<InstalledAssetType[]>;
+}
+
+export class PackageServiceImpl implements PackageService {
+  constructor(
+    private readonly internalEsClient: ElasticsearchClient,
+    private readonly internalSoClient: SavedObjectsClient,
+    private readonly logger: Logger
+  ) {}
+
+  public asScoped(request: KibanaRequest) {
+    const preflightCheck = () => {
+      if (!checkSuperuser(request)) {
+        throw new FleetUnauthorizedError(
+          `User does not have adequate permissions to access Fleet agents.`
+        );
+      }
+    };
+
+    return new PackageClientImpl(
+      this.internalEsClient,
+      this.internalSoClient,
+      this.logger,
+      preflightCheck
+    );
+  }
+
+  public get asInternalUser() {
+    return new PackageClientImpl(this.internalEsClient, this.internalSoClient, this.logger);
+  }
+}
+
+class PackageClientImpl implements PackageClient {
+  constructor(
+    private readonly esClient: ElasticsearchClient,
+    private readonly soClient: SavedObjectsClient,
+    private readonly logger: Logger,
+    private readonly preflightCheck?: () => void | Promise<void>
+  ) {}
+
+  public async fetchFindLatestPackage(packageName: string) {
+    await this.#runPreflight();
+    return fetchFindLatestPackage(packageName);
+  }
+
+  public async getRegistryPackage(packageName: string, packageVersion: string) {
+    await this.#runPreflight();
+    return getRegistryPackage(packageName, packageVersion);
+  }
+
+  public async reinstallEsAssets(
+    assetType: ElasticsearchAssetType,
```

Comment:
thoughts on this? could also just remove this and put it on the consumer to pre-filter the paths (I've only looked at transforms but I'm assuming the other assets also use path for this?) so it doesn't run reinstalls on assets they don't want.

I do like that this is more self documenting vs leaving it to the consumer to know you need to pre-filter.

Also considering changing it to `assetTypes: ElasticsearchAssetType[]`.